### PR TITLE
Feature : Polling blockHandler and intialization handler

### LINKED ISSUE #959: Add a polling block handler
Currently, we have blockHandlers with and without filter. Additionally, it would be great to have a polling filter that runs every nth block (or time window).

E.g.

```yml
blockHandlers
  - handler: handleFundShutDown
    every: 24h
```

### PR DESCRIPTION
**Pull request by incrypto32**
Closes #959, #3221

### CODE REVIEW

File: chain/ethereum/src/data_source.rs
Comment (leoyvens) [inline chain/ethereum/src/data_source.rs:396]: Seems to me the correct matching logic would be to check that `(trigger.block - self.start_block) % handler.filter.every == 0`.

File: chain/ethereum/src/trigger.rs
Comment (leoyvens) [inline chain/ethereum/src/trigger.rs:300]: In principle, I think `EthereumBlockTriggerType` should not carry the address, and `None` should be returned here. As a block trigger, in principle, should not be associated to any specific address since it represents the whole block. `EthereumBlockTriggerType::WithCallTo` is an exception to this principle, and I'd prefer not to add another one.

File: chain/ethereum/src/ethereum_adapter.rs
Comment (incrypto32) [inline chain/ethereum/src/ethereum_adapter.rs:1726]: NOTE: This seems that it can be further optimized. These once triggers are only triggered at the startBlock so this seems very unncessary to send Every triggers in every block

File: chain/ethereum/src/adapter.rs
Comment (incrypto32) [inline chain/ethereum/src/adapter.rs:161]: @leoyvens can something be done here for `to_firehose_filter`, can we optimize the firehose filters based on polling intervals if `trigger_on_every_block` is false and only certain blocks are required?

Here it seems `send_all_block_headers` is only true when trigger_on_every_block is true. Does that mean this feature wont work with firehose?

```
 let combined_filter = CombinedFilter {
            log_filters,
            call_filters,
            send_all_block_headers: trigger_every_block,
        };
 ```

Comment (leoyvens) [reply chain/ethereum/src/adapter.rs:161]: Thanks for raising attention this point, I don't think we could optimize it without adding support for this on the Firehose side.

It does look like the current code would have a bug when syncing with Firehose that would result in missed block triggers. If any `polling_intervals` exist, then we should set `send_all_block_headers` in the Firehose filter.

File: chain/ethereum/src/adapter.rs
Comment (leoyvens) [inline chain/ethereum/src/adapter.rs:164]: We should not use `..` here, as it makes it easy to add a field in the future without considering the consequences to this fn.

File: chain/ethereum/src/adapter.rs
Comment (leoyvens) [inline chain/ethereum/src/adapter.rs:589]: I don't think this needs to carry the `Address`.

File: chain/ethereum/src/data_source.rs
Comment (leoyvens) [inline chain/ethereum/src/data_source.rs:284]: We currently allow having a call filter and an unfiltered block handler in the same ds. So this would be a breaking change, I think we can continue to allow that combination.

But do we really want to allow combining polling and initialization handlers?

File: chain/ethereum/src/data_source.rs
Comment (leoyvens) [inline chain/ethereum/src/data_source.rs:427]: This is very well noticed, and we need to fix before merging. There are two places where we set `start_block: 0` for dynamic data sources, maybe we should just set that to `start_block: creation_block` and this code would work fine.

File: chain/ethereum/src/data_source.rs
Comment (leoyvens) [inline chain/ethereum/src/data_source.rs:433]: Considering we now have the `Once` filter, we shouldn't need to set this to `0` for triggering only once. One cool thing you could try is using a [NonZeroU32](https://doc.rust-lang.org/stable/std/num/struct.NonZeroU32.html) to indicate that this is never zero.

File: chain/ethereum/src/ethereum_adapter.rs
Comment (leoyvens) [inline chain/ethereum/src/ethereum_adapter.rs:1412]: When `trigger_every_block` is set, this is now calling `block_range_to_ptrs` twice, once for Start triggers and then again for End triggers, which is very inefficient. Lets call it only once, and use the result both for the Start and End triggers.

**Review by leoyvens — CHANGES_REQUESTED** [review]

File: chain/ethereum/src/adapter.rs
Comment (maoueh) [reply chain/ethereum/src/adapter.rs:161]: Only possibility I could see would be to use the `HeaderOnly` filter, if you only need the `BlockHeader`, it's a good idea to use to avoid the network transfer of many useless bytes.

But like Leo said, Firehose would need to be adapted to send block each X block.

File: chain/ethereum/src/data_source.rs
Comment (incrypto32) [reply chain/ethereum/src/data_source.rs:284]: > We currently allow having a call filter and an unfiltered block handler in the same ds

@leoyvens do people actually use that combination? I don't see the point in having any filtered handler when there is a non-filtered block handler. is it only because its a breaking change that we want to allow that?

Comment (leoyvens) [reply chain/ethereum/src/data_source.rs:284]: If we can use some tooling to determine that no subgraph uses this combination, we could disallow it. But unless we can prove otherwise, [this usually happens](https://xkcd.com/1172/).

File: chain/ethereum/src/adapter.rs
Comment (incrypto32) [reply chain/ethereum/src/adapter.rs]: I've modified this to set `send_all_block_headers` true when `polling_intervals` are not empty

File: chain/ethereum/src/data_source.rs
Comment (incrypto32) [reply chain/ethereum/src/data_source.rs]: @leoyvens done!

**Review by leoyvens — APPROVED** [review]
Approved! We can merge after an integration test run

**Comment by incrypto32** [comment]
Rebased with master

**Pull request merged**
